File: leetcode/java/0283_move_zeroes.java
class Solution {
    public void moveZeroes(int[] nums) {
        if (nums.length == 1) return;

        int i = 0, j = 1;
        while (j < nums.length) {
            if (nums[i] == 0 && nums[j] != 0) {
                int numJ = nums[j];
                nums[j] = nums[i];
                nums[i] = numJ;
                i++;
            } else if (nums[i] != 0) i++;
            j++;
        }
    }
}

File: leetcode/cpp/0283_move_zeroes.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        // Have one for loop that iterates through nums, with i
        // have a second iterator that increments only when the current nums[j] != 0
        // if nums[j] == 0 && nums[i] != 0, swap their values
        // exit the loop when i == nums.size()
        
        for (int i = 1, j = 0; i < nums.size(); i++) {
            if (nums[j] == 0 && nums[i] != 0) {
                int tmp = nums[j];
                nums[j] = nums[i];
                nums[i] = tmp;
            }
            
            if (nums[j] != 0) {
                j++;
            }
        }
    }
};